drivers/soundwire/intel.c

```diff
@@ -1502,72 +1475,86 @@ static int intel_resume(struct device *dev)
 			pm_runtime_idle(&md->dev);
 	}
 
-	ret = intel_init(sdw);
+	if (clock_stop) {
+		resume = sdw_cdns_check_resume_status(&sdw->cdns);
+		if (resume)
+			return 0; /* it's already running */
+
+		/* Invoke shim for wake disable */
+		intel_shim_wake(sdw, false);
+	}
+
+	ret = intel_init(sdw, clock_stop);
 	if (ret) {
 		dev_err(dev, "%s failed: %d", __func__, ret);
 		return ret;
 	}
 
-	/* make sure all Slaves are tagged as UNATTACHED */
-	sdw_clear_slave_status(&sdw->cdns.bus);
+	if (clock_stop)
+		/* make sure all Slaves are tagged as UNATTACHED */
+		sdw_clear_slave_status(&sdw->cdns.bus);
 
 	ret = sdw_cdns_enable_interrupt(cdns, true);
 	if (ret < 0) {
 		dev_err(dev, "cannot enable interrupts during resume\n");
 		return ret;
 	}
 
-	ret = sdw_cdns_exit_reset(cdns);
-	if (ret < 0) {
-		dev_err(dev, "unable to exit bus reset sequence during resume\n");
-		return ret;
+	if (clock_stop) {
+		ret = sdw_bus_exit_clk_stop(&sdw->cdns.bus);
+
+		if (ret < 0)
+			dev_err(dev, "unable to clock stop during resume\n");
+	} else {
+		ret = sdw_cdns_exit_reset(cdns);
+		if (ret < 0) {
+			dev_err(dev, "unable to exit bus reset sequence during resume\n");
+			return ret;
+		}
 	}
 
 	dev_err(dev, "%s done\n", __func__);
 
 	return ret;
 }
 
-static int intel_resume_runtime(struct device *dev)
+static int intel_suspend(struct device *dev)
+{
+	return _suspend(dev, false);
+}
+
+static int intel_suspend_runtime(struct device *dev)
 {
 	struct sdw_cdns *cdns = dev_get_drvdata(dev);
 	struct sdw_intel *sdw = cdns_to_intel(cdns);
-	int ret;
-
-	if (cdns->bus.prop.hw_disabled) {
-		dev_dbg(dev, "SoundWire master %d is disabled, ignoring\n",
-			cdns->bus.link_id);
-		return 0;
-	}
-
-	dev_err(dev, "%s start\n", __func__);
+	int link_flags;
+	bool clock_stop = true;
 
-	ret = intel_init(sdw);
-	if (ret) {
-		dev_err(dev, "%s failed: %d", __func__, ret);
-		return ret;
-	}
+	link_flags = md_flags >> (sdw->cdns.bus.link_id * 8);
+	if (link_flags & SDW_INTEL_MASTER_DISABLE_CLOCK_STOP)
+		clock_stop = false;
 
-	/* make sure all Slaves are tagged as UNATTACHED */
-	sdw_clear_slave_status(&sdw->cdns.bus);
+	return _suspend(dev, clock_stop);
```

Comment:
I suggest we first work through the code changes for suspend and pm_runtime, and at a later time we merge the code in helpers
This just too hard to review.

Reply:
yes, I will refine it